jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,419 @@
+                    V tail = path.getTail();
+
+                    // the path builds a valid negative cycle
+                    if (graph.containsEdge(tail, head)
+                            && path.getCost() + graph.getEdgeWeight(graph.getEdge(tail, head)) < 0) {
+                        LabeledPath<V> cycleResult = path.clone();
+                        cycleResult.addVertex(head, graph.getEdgeWeight(graph.getEdge(tail, head)), labelMap.get(head));
+
+                        return new GraphWalk<>(graph, cycleResult.getVertices(), cycleResult.getCost());
+                    }
+
+                    for (E e : graph.outgoingEdgesOf(tail)) {
+                        V currentVertex = graph.getEdgeTarget(e);
+                        // extend the path if the extension is still negative and correctly labeled
+                        double edgeWeight = graph.getEdgeWeight(e);
+                        int currentLabel = labelMap.get(currentVertex);
+                        if (!path.getLabels().contains(currentLabel) && path.getCost() + edgeWeight < 0) {
+                            LabeledPath<V> newPath = path.clone();
+                            newPath.addVertex(currentVertex, edgeWeight, currentLabel);
+
+                            /*
+                             * check if paths are dominated, i.e. if the path is definitely worse than other paths
+                             * and does not have to be considered in the future
+                             */
+                            if (!removeDominatedPathsOfLengthKplus1(newPath, pathsLengthKplus1)) {
+                                if(!removeDominatedPathsOfLengthK(newPath, pathsLengthK)) {
+                                    addPathToPathStructure(pathsLengthKplus1, newPath);
+                                }
+                            }
+                        }
+                    }
+
+            }
+            // update k and the corresponding sets
+            k += 1;
+            pathsLengthK = pathsLengthKplus1;
+            pathsLengthKplus1 = new LinkedHashMap<>();
+        }
+
+        return new GraphWalk<>(graph, new ArrayList<>(lengthBound), Double.MAX_VALUE);
+    }
+
+    /**
+     * Checks whether <code>path</code> dominates the current minimal cost path with the same head, tail and label set
+     * in the set of all paths of length k + 1. Thus, dominated paths are eliminated.
+     * This is important out of efficiency reasons, otherwise many unnecessary paths may
+     * be considered in further calculations.
+     *
+     * @param path the currently calculated path
+     * @param pathsLengthKplus1 all before calculated paths of length k + 1
+     *
+     * @return whether <code>path</code> dominates the current minimal cost path with the same head, tail and label set.
+     */
+    private boolean removeDominatedPathsOfLengthKplus1(LabeledPath<V> path, Map<PathSetKey<V>, LabeledPath<V>> pathsLengthKplus1) {
```

Comment:
Since this method no longer does the actual removal, it should be renamed from "remove" to "check".